text/0008-partial-schema-validation.md

```diff
@@ -0,0 +1,207 @@
+After fixing the error, we are left with a policy that is accepted by partial schema validation even when there is no schema available.
+
+```
+permit(principal, action, resource) when {
+  principal.access_level > 5
+};
+```
+
+The policy now passes partial schema validation because `access_level` might exist for the principal entity and it might be a value of type `Long`,
+but we do not *know* that both of these are always true.
+When either of these is not the case, there will be an error during evaluation that partial schema validation cannot detect ahead of time.
+
+### Implementation: Bottom Type
+
+We implement typechecking for an access to an undeclared attribute by assigning all undeclared attributes the bottom type.
+The bottom type functions like the [TypeScript `any` type](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#any).
+It is used to ensure that the type is allowed to appear in any expression without causing additional type errors.
+The `>` operator expects a value of type `Long` as its operand.
+It will in fact accept any subtype of `Long`, and the bottom type is a subtype of every type, so it accepted.
+
+By using the bottom type, we admit some inconsistent expressions.
+For example, `principal.access_level > 5 && principal.access_level like "admin*"` is inconsistent because it requires that the attribute is both a `String` and a `Long`.
+This is not possible for any Cedar value, but the bottom type is nonetheless accepted in both locations.
+
+Importantly, this problem only exists for partial schema validation.
+The standard validation mode does not assign the bottom type to undefined attributes, instead reporting an error immediately.
+
+### Partial Attribute Information
+
+The example so far demonstrates validation without a schema,
+but partial schema validation can use information from a partial schema to detect further errors.
+Suppose we know `principal` is a `User` entity and, for this entity type, `access_level` is not a `Long` but instead is a `String`.
+We could add the following entity type entry to the schema.
+
+```
+"User": {
+  "shape": {
+    "type": "Record",
+    "attributes": {
+      "access_level": {"type": "String"}
+    },
+    "additionalAttributes": true
+  }
+}
+```
+
+We can now typecheck the policy under the assumption that `access_level` is a `String`,
+so the partial schema validator will report an error -
+this time for the left `>` operand.
+Including `"additionalAttributes": true` in the schema informs the validator that the listed attributes are still incomplete,
+so an expression `principal.role == "Admin"` would still be accepted by the validator.
+
+Alternatively, we might have complete information about the attributes present for `User` entities.
+When we know that `access_level` does not exist, the `shape` for the principal entity type can include `"additionalAttributes": false` to specify that there are no undeclared attributes for this entity type.
```

Comment:
Note that `"additionalAttributes": false` is the default, so it is unnecessary to say this explicitly.

Reply:
I see now that it's noted in a later section, but still good to include here to head off confusion.